### ISSUE TITLE
Add typescript index.d.ts file (sparked from Question: How to use it in typescript?)

### ISSUE THREAD

**Issue by Jeff-Tian** [mrodrig/json-2-csv#101]

**Comment by mrodrig** [comment]
Hi @Jeff-Tian. Great question! I'm personally not too familiar with TypeScript, but based on what I found in the TS documentation, it looks like the code should be pretty similar to the "plain" JavaScript examples that I have on the Wiki.  I tried it out and only had to make one change with regards to how the module was imported. In particular, this code is compiling and running successfully with the latest TypeScript compiler:

*test.ts*:
```javascript
import converter = require('json-2-csv');

let documents = [
    {
        Make: 'Nissan',
        Model: 'Murano',
        Year: '2013',
        Specifications: {
            Mileage: '7106',
            Trim: 'S AWD'
        }
    },
    {
        Make: 'BMW',
        Model: 'X5',
        Year: '2014',
        Specifications: {
            Mileage: '3287',
            Trim: 'M'
        }
    }
];

let json2csvCallback = function (err, csv) {
    if (err) throw err;
    console.log(csv);
};

converter.json2csv(documents, json2csvCallback);
```

Compiled with:
```
$ tsc test.ts
```

Compiled file (test.js):
```javascript
"use strict";
exports.__esModule = true;
var converter = require("json-2-csv");
var documents = [
    {
        Make: 'Nissan',
        Model: 'Murano',
        Year: '2013',
        Specifications: {
            Mileage: '7106',
            Trim: 'S AWD'
        }
    },
    {
        Make: 'BMW',
        Model: 'X5',
        Year: '2014',
        Specifications: {
            Mileage: '3287',
            Trim: 'M'
        }
    }
];
var json2csvCallback = function (err, csv) {
    if (err)
        throw err;
    console.log(csv);
};
converter.json2csv(documents, json2csvCallback);
```

I'll add a page to the Wiki with this example too for future reference.

Hope this helps!

**Comment by Jeff-Tian** [comment]
Thank you @mrodrig, works like a charm! Another question: How to use async/await style? I can do it with 

```
const csvData = await new Promise((resolve, reject) => converter.json2csv(data, (err, csv) => {
          if (err) {
            reject(err);
          } else {
            resolve(csv);
          }
        }));
```

Is there any other better suggestion?

**Comment by mrodrig** [comment]
Perfect, glad to hear that worked @Jeff-Tian!

Depending on how you might be using it, I was able to get the async/await style working using the json2csvAsync function which returns a Promise:

```typescript
async function genCsv(data) {
    const csvData = await converter.json2csvAsync(data);
    console.log(data);
    return csvData;
}

genCsv(documents);
```

**Comment by Jeff-Tian** [comment]
Thank you @mrodrig.

**Comment by Ricki-BumbleDev** [comment]
> ```js
> import converter = require('json-2-csv');
> ```

Actually this is the hacky way of importing an untyped package into a TypeScript file. The preferable solution to this issue would be to add an [`index.d.ts` file](http://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html) with the type definitions for this package.

Then one could import the package like so:
```js
import * as json2csv from 'json-2-csv';
```

**Comment by mrodrig** [comment]
Hi @Ricki-BumbleDev! Thanks for commenting on this with a better solution. Since I haven't used TypeScript before I was going off what I found on StackOverflow, so I definitely appreciate you commenting a better approach! I'll update the wiki page with that info too.

Thanks again! Have a great weekend!

**Comment by Ricki-BumbleDev** [comment]
Hi @mrodrig! Thanks for your quick response! I think we've got a bit of a misunterstanding here. If everyone using the package with TypeScript writes his or her own type definition file, it works, however, what I actually meant is adding an `index.d.ts` file to this repository (alternatively to DefinitelyTyped, then the types would be available on npm as `@types/json-2-csv`). Maybe someone finds some time somewhen to write and PR these definitions. I would do it myself, but don't really have the time at the moment.

**Comment by mrodrig** [comment]
Hi @Ricki-BumbleDev! Oh, my bad. I definitely misunderstood. Sorry for my confusion! I found a few references that seem like they might be helpful for someone like myself with no TypeScript experience to go through and setup the `index.d.ts` file like you mentioned, so I might give it a try when I have a chance (unless someone beats me to it and opens a PR 😄).

Here are the links that I found, for my future reference:
https://github.com/Microsoft/TypeScript/issues/8335
https://stackoverflow.com/questions/41891795/how-to-write-a-definition-file-for-commonjs-module-that-exports-function?answertab=votes#tab-top
https://stackoverflow.com/questions/24029462/how-to-write-a-typescript-definition-file-for-a-node-module-that-exports-a-funct
https://www.detroitlabs.com/blog/2018/02/28/adding-custom-type-definitions-to-a-third-party-library/

I'm going to re-open this issue to track adding that `index.d.ts` file.
